pkg/queue/workers/retention.go

```diff
@@ -0,0 +1,204 @@
+	defer func() {
+		h.FinishSpan(span, err)
+	}()
+	span.SetTag("task.id", task.ID)
+	span.SetTag("task.queue", task.Queue)
+	span.SetTag("task.type", task.Type)
+	span.SetTag("task.spec", string(task.Spec))
+
+	log := logrus.WithContext(ctx).WithField("type", task.Type)
+	log.Debug("starting retention task")
+
+	var progress retentionProgress
+	defer func() {
+		// we check for errSerializingHearbeat so we don't cause
+		// a recursion call
+		if err == nil || err == errSerializingHearbeat {
+			return
+		}
+		message := err.Error()
+		progress.ErrorMessage = &message
+		_ = sendRetentionProgress(progress, heartbeats)
+	}()
+
+	spec := retentionSpec{}
+	err = json.Unmarshal(task.Spec, &spec)
+	if err != nil {
+		return fmt.Errorf("can not parse the retention task spec %w", err)
+	}
+
+	span.SetTag("spec.queue", spec.QueueName)
+	span.SetTag("spec.taskType", spec.TaskType)
+	span.SetTag("spec.status", spec.Status)
+	span.SetTag("spec.age", spec.Age.String())
+
+	// initial heartbeat
+	err = sendRetentionProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	result, err := h.db.ExecContext(ctx, spec.SQL)
+	duration := time.Since(now).Milliseconds()
+	progress.Duration = &duration
+
+	if err != nil {
+		return fmt.Errorf("retention execution failed: %w", err)
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("can not get rows affected: %w", err)
+	}
+
+	progress.RowsAffected = &rowsAffected
+	err = sendRetentionProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	log.WithField("rowCount", rowsAffected).Debug("retention finished")
+	return nil
+}
+
+func sendRetentionProgress(progress retentionProgress, heartbeats chan<- queue.Progress) (err error) {
+	logrus.
+		WithField("method", "sendRetentionProgress").
+		Debugf("%+v", progress)
+
+	bytes, err := json.Marshal(progress)
+	if err != nil {
+		logrus.Error(err)
+		return errSerializingHearbeat
+	}
+
+	heartbeats <- bytes
+	return nil
+}
+
+// AssertRetentionSchedule creates a new queue retention tasks for the supplied queue, finished tasks matching
+// the supplied parameters will be deleted
+func AssertRetentionSchedule(ctx context.Context, scheduler queue.Scheduler, queueName string, taskType queue.TaskType, status queue.TaskStatus, filter squirrel.Sqlizer, age time.Duration) error {
+	spec := createRetentionSpec(queueName, taskType, status, filter, age)
+	specBytes, err := json.Marshal(spec)
+	if err != nil {
+		return fmt.Errorf("can not build retention task spec: %w", err)
+	}
+	// randomly distribute the retention tasks throughout the hour
+	when := rand.Intn(60)
+	retentionSchedule := queue.TaskScheduleRequest{
+		TaskBase: queue.TaskBase{
+			Queue: MaintenanceTaskQueue,
+			Type:  RetentionTask,
+			Spec:  specBytes,
+		},
+		CronSchedule: fmt.Sprintf("%d * * * *", when), // every hour at minute "when"
+	}
+
+	return scheduler.AssertSchedule(ctx, retentionSchedule)
```

Comment:
This would not work because the cron schedule is random, it will most likely create a new schedule. I think we should drop this idea here and control it manually in a service where we create the schedule and put the minute we like.

Reply:
I don't want to define these in the service because _all_ of the queues need this retention task, it really is/should be generic to me. I thought that the worst case was that AssertSchedule would update/replace the existing cron?

Reply:
it will not update anything at all. It checks existence and creates if does not exist. However, I checked and the cron schedule is not involved into the existence check at all, so it would be safe to do this but it would set the cron expression only on the first creation.

https://github.com/contiamo/go-base/blob/8dccb682c6d9abce3efcdbb316edb001d5713b8d/pkg/queue/postgres/scheduler.go#L126-L132